### PR TITLE
Phase 1 v2 models foundation expand

### LINKED ISSUE #10: 1.1: Create Zod Validation Schemas (V2)
## Task
Create comprehensive Zod validation schemas for all V2 device and reading operations.

## Requirements
- Create `/lib/validations/v2/device.validation.ts` with schemas for:
  - Device creation (POST)
  - Device updates (PATCH)
  - Query parameters (list/filter devices)
- Create `/lib/validations/v2/reading.validation.ts` with schemas for:
  - Reading creation/ingest
  - Reading queries with filters
  - Bulk insert operations
- Create `/lib/validations/common.validation.ts` with shared utilities:
  - Pagination (cursor, offset, limit)
  - Date ranges
  - Sort orders
  - Common field validations

## Validation Rules
- serial_number: string, required, unique
- manufacturer: string, required
- model: string, required
- firmware_version: string, required
- configuration.threshold_warning: number, required
- configuration.threshold_critical: number, required
- status: enum ['active', 'maintenance', 'offline', 'decommissioned', 'error']
- health.battery_level: number, optional, 0-100 range
- health.signal_strength: number, optional
- reading.value: number, required
- reading.timestamp: Date, required, not in future

## Acceptance Criteria
- [x] All schemas are exported and ready for use
- [x] Input validation tests pass
- [x] Error messages are clear and actionable
- [x] Schemas handle edge cases (negative numbers, out-of-range values)

### LINKED ISSUE #11: 1.2: Create DeviceV2 Model
## Task
Create enhanced DeviceV2 model with production-ready fields in new `devices_v2` collection.

## Model Structure
```typescript
{
  serial_number: string (unique, indexed)
  manufacturer: string
  model: string
  firmware_version: string
  
  configuration: {
    threshold_warning: number
    threshold_critical: number
    sampling_interval: number (seconds)
    calibration_date: Date
    calibration_offset: number
  }
  
  location: {
    building_id: string
    floor: number
    room_name: string
    coordinates?: { x: number, y: number, z: number }
    zone?: string
  }
  
  metadata: {
    tags: string[]
    department: string
    cost_center?: string
    warranty_expiry?: Date
    last_maintenance?: Date
    next_maintenance?: Date
  }
  
  audit: {
    created_at: Date
    created_by: string = "sys-migration-agent"
    updated_at: Date
    updated_by: string
    deleted_at?: Date (soft delete)
    deleted_by?: string
  }
  
  health: {
    last_seen: Date
    uptime_percentage: number
    error_count: number
    last_error?: { timestamp: Date, message: string, code: string }
    battery_level?: number
    signal_strength?: number
  }
  
  status: 'active' | 'maintenance' | 'offline' | 'decommissioned' | 'error'
  status_reason?: string
  
  compliance: {
    requires_encryption: boolean
    data_classification: 'public' | 'internal' | 'confidential' | 'restricted'
    retention_days: number = 90
  }
}
```

## Requirements
- Create `/models/v2/DeviceV2.ts` Mongoose schema and model
- Collection name: `devices_v2`
- Set default values: null for numbers, "unknown" for strings
- Add timestamps (createdAt, updatedAt via audit fields)
- Add soft delete support (check deleted_at field)

## Acceptance Criteria
- [ ] Model exports successfully
- [ ] Collection is `devices_v2`
- [ ] Unique index on serial_number works
- [ ] All required fields have defaults as per decision
- [ ] TypeScript interfaces are properly typed

### LINKED ISSUE #12: 1.3: Create ReadingV2 Model (Timeseries)
## Task
Create enhanced ReadingV2 model optimized for timeseries data in new `readings_v2` collection.

## Model Structure (MongoDB Timeseries)
```typescript
{
  // 1. METADATA (Low cardinality bucketing key)
  metadata: {
    device_id: string (required, indexed)
    type: string (required) // e.g., 'temperature', 'humidity', 'co2'
    unit: string (required) // e.g., 'celsius', 'percent', 'ppm'
    source: string = 'sensor' // enum: ['sensor', 'simulation', 'manual', 'calibration']
  }
  
  // 2. MEASUREMENT DATA
  timestamp: Date (required, indexed)
  value: number (required)
  
  // 3. QUALITY & TRUST
  quality: {
    is_valid: boolean = true
    confidence_score: number (0-1)
    validation_flags: string[]
    is_anomaly: boolean = false
    anomaly_score: number
  }
  
  // 4. CONTEXT
  context: {
    battery_level: number
    signal_strength: number
    ambient_temp: number
  }
  
  // 5. AUDIT (Lightweight)
  processing: {
    raw_value: number
    calibration_offset: number
    ingested_at: Date
  }
}
```

## Requirements
- Create `/models/v2/ReadingV2.ts` Mongoose schema
- Collection name: `readings_v2`
- Use MongoDB timeseries schema options:
  - timeField: 'timestamp'
  - metaField: 'metadata'
  - granularity: 'seconds'
- Set TTL: 90 days (7,776,000 seconds)
- Create compound index: metadata.device_id + timestamp

## Acceptance Criteria
- [ ] Model uses MongoDB timeseries features
- [ ] TTL is set to 90 days
- [ ] Compound index on (device_id, timestamp) exists
- [ ] Default values match decision (null for numbers, "unknown" for strings)
- [ ] ReadingV2 model and collection created successfully

### LINKED ISSUE #13: 1.4: Create Database Indexes (V2)
## Task
Create database indexes for v2 collections to optimize query performance.

### DeviceV2 Indexes
- `serial_number` (unique, ascending)
- `location.building_id + location.floor` (compound)
- `status` (ascending)
- `health.last_seen` (descending - for offline detection)
- `audit.deleted_at` (ascending, sparse - for soft delete queries)

### ReadingV2 Indexes
- `metadata.device_id + timestamp` (compound, required for timeseries)
- `quality.is_anomaly` (ascending)
- `metadata.source` (ascending)

## Requirements
- Create `/scripts/v2/create-indexes-v2.ts`
- Script should be idempotent (safe to run multiple times)
- Add logging for index creation progress
- Handle errors gracefully
- Make executable: `pnpm run create-indexes-v2`

## Index Benefits
- serial_number: Enforces uniqueness, speeds up device lookups
- building_id + floor: Optimizes floor plan queries
- status: Filters devices by operational status
- last_seen: Finds offline devices efficiently
- deleted_at: Queries for soft-deleted devices
- device_id + timestamp: Critical for timeseries read performance
- is_anomaly: Fast anomaly queries
- source: Filters by data source

## Acceptance Criteria
- [ ] All indexes created successfully
- [ ] Script is idempotent
- [ ] Index names are clear and descriptive
- [ ] Run time is reasonable (< 30 seconds)
- [ ] Sparse indexes use sparse: true option

### LINKED ISSUE #14: 1.5: Create TypeScript Type Definitions (V2)
## Task
Create TypeScript type definitions for V2 models and API types.

### `/types/v2/device.types.ts`
- DeviceV2Document (Mongoose document)
- DeviceV2Input (for creation)
- DeviceV2UpdateInput (for updates)
- DeviceConfig, DeviceLocation, DeviceMetadata, etc.
- DeviceStatus enum
- DeviceHealthMetrics interface

### `/types/v2/reading.types.ts`
- ReadingV2Document (Mongoose document)
- ReadingV2Input (for creation/ingest)
- ReadingMetadata, ReadingQuality, ReadingContext, ReadingProcessing interfaces
- ReadingSource enum

### `/types/v2/api.types.ts`
- ListDevicesQuery (query parameters)
- ListDevicesResponse with pagination
- GetDeviceResponse
- CreateDeviceRequest/Response
- UpdateDeviceRequest/Response
- ListReadingsQuery
- IngestReadingsRequest/Response
- AnalyticsQuery and response types
- ErrorResponse with error codes

## Requirements
- Export all types from `/types/v2/index.ts` for easy imports
- Types should match Zod schemas from Phase 1.1
- Use discriminated unions where appropriate
- Include optional fields properly
- Document complex types with JSDoc comments

## Acceptance Criteria
- [ ] All types compile without errors
- [ ] Types match model definitions
- [ ] Zod schemas align with TypeScript types
- [ ] Types are exported from `/types/v2/index.ts`
- [ ] No implicit any types

### LINKED ISSUE #15: 1.6: Create Migration Utilities
## Task
Create migration utilities to support dual-write during transition period (optional but useful for testing).

### `/lib/migration/dual-write-adapter.ts`
- Function: `writeToBothCollections(data, operation)`
- Attempts to write to both v1 and v2 collections
- Logs failures in v2 but doesn't fail entire operation if v2 write fails
- Supports: create, update, delete operations
- Add retry logic with exponential backoff

### `/lib/migration/v1-to-v2-mapper.ts`
- Function: `mapDeviceV1toV2(v1Device): DeviceV2`
- Function: `mapReadingV1toV2(v1Reading): ReadingV2`
- Handles default values: null for numbers, "unknown" for strings
- Maps timestamps to audit.created_at
- Sets created_by: "sys-migration-agent"
- Maps v1 fields to v2 structure

### `/scripts/v2/backfill-v2.ts`
- Script: One-time migration from v1 to v2 collections
- Uses mapper to convert documents
- Reads from v1 collections, writes to v2
- Supports batching and progress tracking
- Idempotent (safe to run multiple times)
- Make executable: `pnpm run backfill-v2`

## Requirements
- Dual-write adapter should be non-blocking (async)
- Mapper functions should be pure and testable
- Backfill script should log progress every 1000 documents
- Handle errors gracefully without stopping migration

## Acceptance Criteria
- [ ] Dual-write adapter works with all operation types
- [ ] Mapper correctly transforms v1 to v2 data
- [ ] Backfill script is idempotent
- [ ] Progress logging shows migration status
- [ ] No data is lost or duplicated during migration

### PR DESCRIPTION
**Pull request by Amadou-dot**
This pull request introduces several enhancements and updates focused on supporting the V2 migration, improving developer experience, and updating dependencies for the Infrasight IoT monitoring dashboard. Key changes include new documentation for architecture and workflows, new scripts for V2 migration, V2 models and validation schemas, and dependency upgrades. Additionally, some minor codebase cleanups and workflow adjustments are included.

**Migration & V2 Support**

* Added V2 MongoDB models (`DeviceV2`, `ReadingV2`) with expanded audit, health, and compliance fields in `models/v2/index.ts`, supporting the Expand-Contract migration pattern.
* Introduced comprehensive Zod validation schemas for V2 API operations in `lib/validations/v2/index.ts`, enabling request validation and TypeScript inference.
* Added new npm scripts in `package.json` for V2 migration tasks: `create-indexes-v2` and `backfill-v2` for setting up indexes and backfilling data.

**Documentation & Developer Experience**

* Added a detailed project overview, architecture, migration plan, code conventions, and troubleshooting guide to `.github/copilot-instructions.md` for onboarding and AI agent support.

**Dependency & Tooling Upgrades**

* Upgraded dependencies in `package.json` and `pnpm-lock.yaml`: `mongoose` to 9.0.1, `recharts` to 3.5.1, `@eslint/eslintrc`, `@tailwindcss/postcss`, `@types/node`, `@types/react`, `@types/react-dom`, `shadcn`, `tailwindcss`, `typescript`, and added `zod` for schema validation. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L23-R48) [[2]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL39-R40) [[3]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL63-R103) [[4]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL283-R287) [[5]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL526-R530) [[6]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL536-R540) [[7]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL903-R907) [[8]](diffhunk://#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL1080-R1084)

**Codebase Maintenance**

* Removed unused direct imports of `Pusher` from several React components (`AnomalyChart.tsx`, `DeviceGrid.tsx`, `FloorPlan.tsx`) to rely on the shared `getPusherClient` utility. [[1]](diffhunk://#diff-e7d6e45a24522d684b5738ace1478070610f589011de16115c59fef2a3591a36L2) [[2]](diffhunk://#diff-5e1a247d0bbcc7efb5c102d57f5effafe96a012a3821de85b4b8ef696b8dc748L2) [[3]](diffhunk://#diff-a1dd7c5bcd418b9316b20fd5a22d3e0f420cea89af6ffe234c504214971e2834L2)
* Minor cleanup in `lib/db.ts` for global mongoose cache declaration.

**CI/CD & Automation**

* Updated the IoT data simulation workflow schedule to run every 5 minutes instead of every minute in `.github/workflows/simulate-data.yml`.

closes #10
closes  #11
closes #12 
closes #13 
closes #14 
closes #15

### CODE REVIEW

File: types/v2/device.types.ts [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+  /** Cost center for billing */
+  cost_center?: string;
+  /** Warranty expiration date */
+  warranty_expiry?: Date;
+  /** Last maintenance date */
+  last_maintenance?: Date;
+  /** Next scheduled maintenance date */
+  next_maintenance?: Date;
+}
+
```
[the remark this answers is not in the document]
Comment (Amadou-dot) [reply types/v2/device.types.ts]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/Amadou-dot/infrasight/pull/34#discussion_r2608526381)
